File: frontend/src/types/job.ts
export type JobSourceType = 'manual' | 'import' | 'donor' | 'api' | 'career_site'
export type JobStatus = 'active' | 'expired' | 'archived'
export type JobSkillImportance = 'required' | 'preferred' | 'mentioned'
export type WorkMode = 'remote' | 'hybrid' | 'onsite' | 'unknown'
export type EmploymentType = 'full_time' | 'part_time' | 'contract' | 'internship' | 'unknown'
export type ApplicationStatus =
  | 'saved'
  | 'preparing'
  | 'applied'
  | 'screening'
  | 'assessment'
  | 'interview'
  | 'offer'
  | 'rejected'
  | 'withdrawn'
  | 'accepted'
  | 'ghosted'
export type ApplicationPriority = 'low' | 'medium' | 'high'
export type IngestionRunStatus = 'running' | 'completed' | 'failed' | 'partial'

export interface JobSkillPublic {
  id: string
  name: string
  slug: string
  importance: JobSkillImportance
}

export interface JobRolePublic {
  id: string
  name: string
  slug: string
  mapping_source: string | null
}

export interface JobCard {
  id: string
  slug: string
  title: string
  company_name: string
  company_slug: string | null
  location_text: string | null
  work_mode: WorkMode | null
  employment_type: EmploymentType | null
  experience_min_years: number | null
  experience_max_years: number | null
  posted_at: string | null
  status: JobStatus
  is_remote: boolean | null
  top_skills: string[]
  is_saved: boolean
}

export interface JobListResponse {
  items: JobCard[]
  total: number
  page: number
  limit: number
}

export interface JobPracticeLink {
  label: string
  path: string
  reason: string | null
}

export interface JobDetail {
  id: string
  slug: string
  title: string
  company_name: string
  company_slug: string | null
  company_id: string | null
  description: string
  requirements_text: string | null
  responsibilities_text: string | null
  employment_type: EmploymentType | null
  work_mode: WorkMode | null
  experience_min_years: number | null
  experience_max_years: number | null
  salary_min: string | null
  salary_max: string | null
  salary_currency: string | null
  location_text: string | null
  country: string | null
  state: string | null
  city: string | null
  source_url: string | null
  apply_url: string | null
  posted_at: string | null
  expires_at: string | null
  status: JobStatus
  is_remote: boolean | null
  source_name: string | null
  skills: JobSkillPublic[]
  roles: JobRolePublic[]
  is_saved: boolean
  application_id: string | null
  application_status: ApplicationStatus | null
  practice_links: JobPracticeLink[]
  interview_prep_url: string | null
  company_prep_url: string | null
}

export interface SavedJobItem {
  id: string
  job_id: string
  saved_at: string
  job: JobCard
}

export interface ApplicationSummary {
  id: string
  job_id: string
  job_title: string
  company_name: string
  status: ApplicationStatus
  applied_at: string | null
  next_follow_up_at: string | null
  priority: ApplicationPriority
  job_status: JobStatus
}

export interface ApplicationDetail {
  id: string
  job_id: string
  status: ApplicationStatus
  applied_at: string | null
  next_follow_up_at: string | null
  source_of_application: string | null
  application_url: string | null
  notes: string | null
  salary_expected: string | null
  priority: ApplicationPriority
  created_at: string
  updated_at: string
  job: JobDetail
}

export interface ApplicationStatusHistoryItem {
  id: string
  from_status: ApplicationStatus | null
  to_status: ApplicationStatus
  note: string | null
  changed_at: string
}

export interface ApplicationUpdate {
  status?: ApplicationStatus
  next_follow_up_at?: string | null
  notes?: string | null
  salary_expected?: string | null
  priority?: ApplicationPriority
  application_url?: string | null
  source_of_application?: string | null
}

export interface ApplicationStatusChange {
  to_status: ApplicationStatus
  note?: string | null
}

export interface JobsSummary {
  saved_count: number
  applications_total: number
  applied_count: number
  interview_count: number
  offer_count: number
  rejected_count: number
  follow_ups_due: number
  follow_ups_today: number
  follow_ups_overdue: number
}

export interface JobPreferenceUpdate {
  target_role_slug?: string | null
  preferred_locations?: string[] | null
  remote_preference?: WorkMode | null
}

export interface JobSourcePublic {
  id: string
  name: string
  slug: string
  source_type: JobSourceType
  is_active: boolean
}

export interface AdminJobCreate {
  title: string
  company_name: string
  description: string
  requirements_text?: string | null
  responsibilities_text?: string | null
  employment_type?: EmploymentType | null
  work_mode?: WorkMode | null
  experience_min_years?: number | null
  experience_max_years?: number | null
  location_text?: string | null
  city?: string | null
  state?: string | null
  country?: string | null
  is_remote?: boolean | null
  source_url?: string | null
  apply_url?: string | null
  posted_at?: string | null
  expires_at?: string | null
  skills?: string[]
  roles?: string[]
  status?: JobStatus
}

export interface AdminJobUpdate {
  title?: string
  company_name?: string
  description?: string
  requirements_text?: string | null
  responsibilities_text?: string | null
  employment_type?: EmploymentType | null
  work_mode?: WorkMode | null
  experience_min_years?: number | null
  experience_max_years?: number | null
  location_text?: string | null
  city?: string | null
  state?: string | null
  country?: string | null
  is_remote?: boolean | null
  source_url?: string | null
  apply_url?: string | null
  posted_at?: string | null
  expires_at?: string | null
  skills?: string[] | null
  roles?: string[] | null
  status?: JobStatus | null
}

export interface ImportPreviewRow {
  row_number: number
  title: string
  company: string
  action: string
  errors: string[]
}

export interface ImportPreviewResponse {
  run_id: string | null
  rows: ImportPreviewRow[]
  valid_count: number
  error_count: number
  create_count: number
  update_count: number
  duplicate_count: number
}

export interface ImportConfirmRequest {
  preview_id: string
  filename?: string | null
}

export interface ImportConfirmResponse {
  run_id: string
  status: IngestionRunStatus
  records_created: number
  records_updated: number
  records_skipped: number
  records_failed: number
}

export interface IngestionRunPublic {
  id: string
  source_id: string | null
  source_name: string | null
  started_at: string
  completed_at: string | null
  status: IngestionRunStatus
  records_seen: number
  records_created: number
  records_updated: number
  records_skipped: number
  records_failed: number
  source_file_name: string | null
}

export interface IngestionErrorPublic {
  id: string
  row_number: number | null
  external_id: string | null
  error_type: string
  message: string
}

export interface AdminJobListResponse {
  items: JobCard[]
  total: number
  page: number
  limit: number
}

export interface JobListFilters {
  q?: string
  role?: string
  skill?: string
  company?: string
  city?: string
  state?: string
  country?: string
  remote?: boolean
  work_mode?: string
  employment_type?: string
  experience_min?: number
  posted_within_days?: number
  sort?: string
  page?: number
  limit?: number
}

File: backend/app/schemas/job.py
"""Pydantic schemas for Build 9 jobs domain."""

from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field

from app.models.job_enums import (
    ApplicationPriority,
    ApplicationStatus,
    EmploymentType,
    IngestionRunStatus,
    JobSkillImportance,
    JobSourceType,
    JobStatus,
    WorkMode,
)


class JobSkillPublic(BaseModel):
    id: UUID
    name: str
    slug: str
    importance: JobSkillImportance


class JobRolePublic(BaseModel):
    id: UUID
    name: str
    slug: str
    mapping_source: str | None = None


class JobCard(BaseModel):
    id: UUID
    slug: str
    title: str
    company_name: str
    company_slug: str | None = None
    location_text: str | None = None
    work_mode: WorkMode | None = None
    employment_type: EmploymentType | None = None
    experience_min_years: int | None = None
    experience_max_years: int | None = None
    posted_at: datetime | None = None
    status: JobStatus
    is_remote: bool | None = None
    top_skills: list[str] = []
    is_saved: bool = False
    requirement_coverage: float | None = None
    has_sufficient_mapping: bool | None = None
    missing_skill_count: int | None = None


class JobListResponse(BaseModel):
    items: list[JobCard]
    total: int
    page: int
    limit: int


class JobPracticeLink(BaseModel):
    label: str
    path: str
    reason: str | None = None


class JobDetail(BaseModel):
    id: UUID
    slug: str
    title: str
    company_name: str
    company_slug: str | None = None
    company_id: UUID | None = None
    description: str
    requirements_text: str | None = None
    responsibilities_text: str | None = None
    employment_type: EmploymentType | None = None
    work_mode: WorkMode | None = None
    experience_min_years: int | None = None
    experience_max_years: int | None = None
    salary_min: Decimal | None = None
    salary_max: Decimal | None = None
    salary_currency: str | None = None
    location_text: str | None = None
    country: str | None = None
    state: str | None = None
    city: str | None = None
    source_url: str | None = None
    apply_url: str | None = None
    posted_at: datetime | None = None
    expires_at: datetime | None = None
    status: JobStatus
    is_remote: bool | None = None
    source_name: str | None = None
    skills: list[JobSkillPublic] = []
    roles: list[JobRolePublic] = []
    is_saved: bool = False
    application_id: UUID | None = None
    application_status: ApplicationStatus | None = None
    practice_links: list[JobPracticeLink] = []
    interview_prep_url: str | None = None
    company_prep_url: str | None = None
    match: dict[str, Any] | None = None


class SavedJobItem(BaseModel):
    id: UUID
    job_id: UUID
    saved_at: datetime
    job: JobCard


class ApplicationSummary(BaseModel):
    id: UUID
    job_id: UUID
    job_title: str
    company_name: str
    status: ApplicationStatus
    applied_at: datetime | None = None
    next_follow_up_at: datetime | None = None
    priority: ApplicationPriority
    job_status: JobStatus


class ApplicationDetail(BaseModel):
    id: UUID
    job_id: UUID
    status: ApplicationStatus
    applied_at: datetime | None = None
    next_follow_up_at: datetime | None = None
    source_of_application: str | None = None
    application_url: str | None = None
    notes: str | None = None
    salary_expected: Decimal | None = None
    priority: ApplicationPriority
    created_at: datetime
    updated_at: datetime
    job: JobDetail


class ApplicationStatusHistoryItem(BaseModel):
    id: UUID
    from_status: ApplicationStatus | None = None
    to_status: ApplicationStatus
    note: str | None = None
    changed_at: datetime


class ApplicationUpdate(BaseModel):
    status: ApplicationStatus | None = None
    next_follow_up_at: datetime | None = None
    notes: str | None = None
    salary_expected: Decimal | None = None
    priority: ApplicationPriority | None = None
    application_url: str | None = None
    source_of_application: str | None = None


class ApplicationStatusChange(BaseModel):
    to_status: ApplicationStatus
    note: str | None = None


class JobsSummary(BaseModel):
    saved_count: int
    applications_total: int
    applied_count: int
    interview_count: int
    offer_count: int
    rejected_count: int
    follow_ups_due: int
    follow_ups_today: int
    follow_ups_overdue: int


class JobPreferenceUpdate(BaseModel):
    target_role_slug: str | None = None
    preferred_locations: list[str] | None = None
    remote_preference: WorkMode | None = None


class JobSourcePublic(BaseModel):
    id: UUID
    name: str
    slug: str
    source_type: JobSourceType
    is_active: bool


class AdminJobCreate(BaseModel):
    title: str = Field(min_length=2, max_length=255)
    company_name: str = Field(min_length=1, max_length=255)
    description: str = Field(min_length=10)
    requirements_text: str | None = None
    responsibilities_text: str | None = None
    employment_type: EmploymentType | None = None
    work_mode: WorkMode | None = None
    experience_min_years: int | None = Field(default=None, ge=0, le=50)
    experience_max_years: int | None = Field(default=None, ge=0, le=50)
    location_text: str | None = None
    city: str | None = None
    state: str | None = None
    country: str | None = None
    is_remote: bool | None = None
    source_url: str | None = None
    apply_url: str | None = None
    posted_at: datetime | None = None
    expires_at: datetime | None = None
    skills: list[str] = []
    roles: list[str] = []
    status: JobStatus = JobStatus.ACTIVE


class AdminJobUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=2, max_length=255)
    company_name: str | None = None
    description: str | None = None
    requirements_text: str | None = None
    responsibilities_text: str | None = None
    employment_type: EmploymentType | None = None
    work_mode: WorkMode | None = None
    experience_min_years: int | None = Field(default=None, ge=0, le=50)
    experience_max_years: int | None = Field(default=None, ge=0, le=50)
    location_text: str | None = None
    city: str | None = None
    state: str | None = None
    country: str | None = None
    is_remote: bool | None = None
    source_url: str | None = None
    apply_url: str | None = None
    posted_at: datetime | None = None
    expires_at: datetime | None = None
    skills: list[str] | None = None
    roles: list[str] | None = None
    status: JobStatus | None = None


class ImportPreviewRow(BaseModel):
    row_number: int
    title: str
    company: str
    action: str
    errors: list[str] = []


class ImportPreviewResponse(BaseModel):
    run_id: UUID | None = None
    rows: list[ImportPreviewRow]
    valid_count: int
    error_count: int
    create_count: int
    update_count: int
    duplicate_count: int


class ImportConfirmResponse(BaseModel):
    run_id: UUID
    status: IngestionRunStatus
    records_created: int
    records_updated: int
    records_skipped: int
    records_failed: int


class IngestionRunPublic(BaseModel):
    id: UUID
    source_id: UUID | None = None
    source_name: str | None = None
    started_at: datetime
    completed_at: datetime | None = None
    status: IngestionRunStatus
    records_seen: int
    records_created: int
    records_updated: int
    records_skipped: int
    records_failed: int
    source_file_name: str | None = None


class ImportConfirmRequest(BaseModel):
    preview_id: UUID
    filename: str | None = None


class IngestionErrorPublic(BaseModel):
    id: UUID
    row_number: int | None = None
    external_id: str | None = None
    error_type: str
    message: str
